solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs

```diff
@@ -295,26 +295,123 @@ impl TokenTransferValidationContext for IbcStorage<'_, '_> {
 
     fn send_coins_validate(
         &self,
-        _from_account: &Self::AccountId,
-        _to_account: &Self::AccountId,
-        _coin: &PrefixedCoin,
+        from_account: &Self::AccountId,
+        to_account: &Self::AccountId,
+        coin: &PrefixedCoin,
     ) -> Result<(), TokenTransferError> {
+        /*
+           Should have the following accounts
+           - token program
+           - token account
+           - escrow account ( with seeds as portId, channelId and denom )
+           - token mint
+
+           If sending tokens from escrow then,
+           - mint authority should be present
+           - from account should match escrow
+           - to account should match token account
+
+          If sending tokens to escrow then,
+           - sender should be present
+           - sender should be signer
+           - to account should match escrow
+           - from account should match token account
+        */
+        let store = self.borrow();
+        let accounts = &store.accounts;
+        if accounts.token_program.is_none() ||
+            accounts.token_mint.is_none() ||
+            accounts.escrow_account.is_none() ||
+            accounts.token_account.is_none()
+        {
+            return Err(TokenTransferError::ParseAccountFailure);
+        }
+        let escrow_account = &accounts.escrow_account.clone().unwrap(); // we check before if it exists so the unwrap is infallible
+        let token_account = &accounts.token_account.clone().unwrap(); // we check before if it exists so the unwrap is infallible
```

Comment:
How about:

```suggestion
        let escrow_account = accounts.escrow_account.as_ref().ok_or(TokenTransferError::ParseAccountFailure)?;
        let token_account = accounts.token_account.as_ref().ok_or(TokenTransferError::ParseAccountFailure)?;
```

This eliminates the need for unwrap and explicit is_none check.
There’s also no need to `clone` the account.

Ditto in `mint_coins_validate` below.